Add more debugging tools for the OCaml run-time and code

stijn-devriendt opened the pull request:
This patch series adds a couple more commands to use when debugging OCaml code and especially the OCaml run-time.

This code is best used with the OCaml debug run-time (and debug symbols), yet it should have been hardened to work without them too, albeit without some of the functionality.
Some improvements could be made in the future (e.g. do pointer arithmetic for structures, rather than depending on the type information being available).

Added commands:
ml_target: similar to the gdb 'target' command, yet uses OCaml-specific information about heap/stacks/... to further annotate the provided information.
ml_find: similar to the gdb 'find' command, but also provides extra annotation
ml_validate: validates the OCaml heap contents. This is only useful with a debug run-time with debug symbols.
ml_show: analogous to ml_scan, validates and shows a single OCaml value. Can recurse down references to other values. If used with debug run-time & debug symbols, more validation is possible.
One TODO for this command is the proper interpretation of lists. Without it, each element requires a level of recursion.

Some of the commands have varying levels of verbosity.

stijn-devriendt committed:
Add and use 2 decorators that display proper exception messages

gdb can't be trusted to display a meaningful backtrace, it seems.

stijn-devriendt committed:
Make the most-used types global for performance and memory footprint

Callsites will be migrated over with further changes

stijn-devriendt committed:
Add memory address space parsing

Parse info from "info target" and "info proc mappings", then annotate with
info obtained through ocaml symbols

stijn-devriendt committed:
Add ml_target command to display ocaml memory address space information

stijn-devriendt committed:
Add ml_find to find all locations with a particular value

stijn-devriendt committed:
factor out symbol resolution

stijn-devriendt committed:
Major overhaul of existing code in preparation of next patches

This change mostly does the following:
- enable python3 (e.g. print/assert are now a function)
- extra hardening against ocaml bugs (e.g. validate string
  size and padding)
- add documentation

stijn-devriendt committed:
Add OCamlValue.try_parse and friends for improved parsing & validation

stijn-devriendt committed:
Add code to find all GC roots

stijn-devriendt committed:
Add ml_validate command to validate ocaml heap

stijn-devriendt committed:
Add ml_show to (recursively) print a value. Final doc updates

ygrek commented:
Wow, thanks a lot!
I will probably merge this straight ahead because I cannot review python code in any sensible way, but let me ask : 
- the modules imported at the beginning - they require some extra packages or they are builtin with standard python installation?
- (looking at print changes) is it python3 compatible now?

stijn-devriendt commented:
You're welcome.

The modules are all standard python modules. I believe the enum module is Python 3 specific (but there's a library available for python 2 support: https://pypi.python.org/pypi/enum/)
And that last bit answers your second question: yes it's python 3 compatible now. That was the first thing we ran into when starting to use your script, our gdb used python 3.

Hopefully I didn't break python2 support...

ygrek commented:
It doesn't work for me sadly on ocaml 4.02.3

With gdb 7.7.1 (using python 2) it gives : 
```
(gdb) ml_dump caml_globals 1
Didn't find 'caml_thread_struct'. System thread roots unavailable.
Traceback (most recent call last):
  File "/home/ygrek/work/p/scraps/mlvalues.py", line 658, in init_memoryspace
    memoryspace = MemorySpace()
  File "/home/ygrek/work/p/scraps/mlvalues.py", line 306, in __init__
    self.populate_ranges()
  File "/home/ygrek/work/p/scraps/mlvalues.py", line 494, in populate_ranges
    start, dash, end, str_is, memtype = line.split(maxsplit=4)
TypeError: split() takes no keyword arguments
Traceback (most recent call last):
  File "/home/ygrek/work/p/scraps/mlvalues.py", line 123, in wrapper
    return f(*args, **kwargs)
  File "/home/ygrek/work/p/scraps/mlvalues.py", line 1362, in invoke
    init_memoryspace()
  File "/home/ygrek/work/p/scraps/mlvalues.py", line 658, in init_memoryspace
    memoryspace = MemorySpace()
  File "/home/ygrek/work/p/scraps/mlvalues.py", line 306, in __init__
    self.populate_ranges()
  File "/home/ygrek/work/p/scraps/mlvalues.py", line 494, in populate_ranges
    start, dash, end, str_is, memtype = line.split(maxsplit=4)
TypeError: split() takes no keyword arguments
Python Exception <type 'exceptions.TypeError'> split() takes no keyword arguments: 
Error occurred in Python command: split() takes no keyword arguments
```

with gdb 7.11.1 (using python 3) : 
```
(gdb) ml_dump caml_globals 1
Didn't find 'caml_thread_struct'. System thread roots unavailable.
Annotating 'major GC's current gray values' over 2 separate ranges: 0x7EFF11254010 - 0x7EFF11294010 is     GC_Metadata|major GC's current gray values and 0x7EFF11294010 - 0x7EFF11454010 is         Unknown|unknown
Traceback (most recent call last):
  File "/home/ygrek/work/p/scraps/mlvalues.py", line 658, in init_memoryspace
    memoryspace = MemorySpace()
  File "/home/ygrek/work/p/scraps/mlvalues.py", line 310, in __init__
    self.annotate_finalisers()
  File "/home/ygrek/work/p/scraps/mlvalues.py", line 637, in annotate_finalisers
    table = gdb.parse_and_eval("final_table").cast(size_t)
gdb.error: No symbol "final_table" in current context.
Traceback (most recent call last):
  File "/home/ygrek/work/p/scraps/mlvalues.py", line 123, in wrapper
    return f(*args, **kwargs)
  File "/home/ygrek/work/p/scraps/mlvalues.py", line 1362, in invoke
    init_memoryspace()
  File "/home/ygrek/work/p/scraps/mlvalues.py", line 658, in init_memoryspace
    memoryspace = MemorySpace()
  File "/home/ygrek/work/p/scraps/mlvalues.py", line 310, in __init__
    self.annotate_finalisers()
  File "/home/ygrek/work/p/scraps/mlvalues.py", line 637, in annotate_finalisers
    table = gdb.parse_and_eval("final_table").cast(size_t)
gdb.error: No symbol "final_table" in current context.
Python Exception <class 'gdb.error'> No symbol "final_table" in current context.: 
Error occurred in Python command: No symbol "final_table" in current context.
```

afaics it breaks in init_memoryspace and doesn't even get to executing ml_dump, can we make it more robust so that missing pieces (in debugged binary?) do not prevent rest of script to run. I am ok to forgo python2 support.

ygrek commented:
Also please put docs at the top of the file describing new commands and short examples of usage.

stijn-devriendt committed:
Improve robustness against missing symbols.

Some hardening was already done (mostly on missing types or specific
variables that were encountered while testing), but apparently more is
needed.

Introduce get_value_safe() which returns None if the provided name doesn't
exist and act upon it by short-circuiting the necessary logic.

stijn-devriendt committed:
mlvalues: Add documentation of new commands at top of file

ygrek commented:
Thanks!
Apparently in my case (gdb 7.11 pythong 3.5) more casts are necessary to make it work, does this look valid? : 
```
diff --git a/mlvalues.py b/mlvalues.py
index 9250d16..75a4220 100644
--- a/mlvalues.py
+++ b/mlvalues.py
@@ -650,7 +650,7 @@ class MemorySpace:
 
         heap_chunk_ptr = heap_chunk_head["next"].cast(heap_chunk_head_p)
     except gdb.MemoryError:
-      print("OCaml major heap linked list is corrupt: last entry = 0x%08X" % (heap_chunk_ptr.cast(size_t)))
+      print("OCaml major heap linked list is corrupt: last entry = 0x%08X" % (int(heap_chunk_ptr.cast(size_t))))
 
     gray_vals = get_value_safe("gray_vals", size_t)
     gray_vals_cur = get_value_safe("gray_vals_cur", size_t)
@@ -711,9 +711,9 @@ def init_memoryspace(reload=False):
 
 def resolve(address):
   """Resolve an address to a symbol (function/variable name)."""
-  symbol = gdb.execute("info symbol 0x%08X" % address.cast(size_t), False, True).split(" ",1)[0]
+  symbol = gdb.execute("info symbol 0x%08X" % int(address.cast(size_t)), False, True).split(" ",1)[0]
   if symbol == "No": # FIXME "No symbol matches"
-    return "0x%08X" % address.cast(size_t)
+    return "0x%08X" % int(address.cast(size_t))
   else:
     return "%s" % symbol
 
@@ -955,7 +955,7 @@ class OCamlValue:
     return resolve(self.val())
 
   def show_opaque(self,s):
-    print("<%s at 0x%x>" % (s,self.val()))
+    print("<%s at 0x%x>" % (s,int(self.val())))
 
   def show_seq(self,seq,delim,recurse,raw=False):
     for i, x in enumerate(seq):
@@ -1398,7 +1398,7 @@ Optional /r flag controls the recursion depth limit."""
       return gdb.parse_and_eval("*((size_t*)&"+addr+")").cast(size_t.pointer())
 
   def show_ptr(self, addr, recurse):
-    print("*0x%x:" % addr.cast(size_t))
+    print("*0x%x:" % int(addr.cast(size_t)))
     OCamlValue(addr.dereference()).show(recurse)
     print("")
 
@@ -1429,7 +1429,7 @@ Optional /r flag controls the recursion depth limit."""
       if args[0] == "local_roots":
         p = gdb.parse_and_eval("*(struct caml__roots_block**)&caml_local_roots")
         while p != 0:
-          print("caml_frame 0x%x" % p.cast(size_t))
+          print("caml_frame 0x%x" % int(p.cast(size_t)))
           for i in range(int(p['nitems'])):
             self.show_ptr(p['tables'][i], recurse)
           p = p['next']
@@ -1520,7 +1520,7 @@ Optional /r flag controls the recursion depth limit."""
       return gdb.parse_and_eval("*((size_t*)&"+addr+")").cast(size_t.pointer())
 
   def show_val(self, addr, recurse):
-    print("0x%x = " % addr.cast(size_t))
+    print("0x%x = " % int(addr.cast(size_t)))
     OCamlValue(addr).show(recurse)
     print("")
```

also now I see `ml_dump` output is broken in more lines than needed, is it possible to print without newlines?
```
(gdb) ml_dump &camlList
(
Closure(
2 fields
)
```

`ml_show` output looks neat, but I still want to keep more dumb `ml_dump` in case smth goes wrong.

stijn-devriendt committed:
Restore single-line behaviour of ml_dump

While replacing print with print(), the statement form's peculiarity of not
appending \n when the statement ends with a comma was overseen.
Introduce print_cont and use it where required.

stijn-devriendt commented:
That patch looks OK to me. I'm surprised we didn't run into that... I'm absolutely in favor of keeping ml_dump, it's served us well, even when ml_show was available.

I have a fix for the single-line ml_dump, but it's a rather simple and naïve one:
```
diff --git a/mlvalues.py b/mlvalues.py
index 9250d16..d7365b1 100644
--- a/mlvalues.py
+++ b/mlvalues.py
@@ -172,6 +172,10 @@ def try_dereference(gdbval):
   except gdb.MemoryError:
     return None
 
+def print_cont(*args, **kwargs):
+  kwargs["end"] = ''
+  print(*args, **kwargs)
+
 # gdb.Type's used often throughout the script
 intnat = size_t = charp = doublep = heap_chunk_head_p = caml_contextp = caml_thread_structp = None
 
@@ -955,14 +959,14 @@ class OCamlValue:
     return resolve(self.val())
 
   def show_opaque(self,s):
-    print("<%s at 0x%x>" % (s,self.val()))
+    print_cont("<%s at 0x%x>" % (s,self.val()))
 
   def show_seq(self,seq,delim,recurse,raw=False):
     for i, x in enumerate(seq):
       if i:
-        print(delim)
+        print_cont(delim)
       if raw:
-        print(x.resolve())
+        print_cont(x.resolve())
       else:
         x.show(recurse)
<snip - rest of patch>
```

If you're ok with it, I'll push it onto the stack.

ygrek commented:
Yes, if I use print_cont instead of print in OCamlValue::show it matches current compact output, so it is fine. Thanks!

stijn-devriendt commented:
I've added the patch. Please let me know if there's any other remarks.

ygrek commented:
Thanks!
